astropy/utils/masked/core.py

```diff
@@ -995,10 +995,80 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
                 # We don't support this.
                 return NotImplemented
 
-        elif method in {"reduceat", "at"}:  # pragma: no cover
-            raise NotImplementedError(
-                "masked instances cannot yet deal with 'reduceat' or 'at'."
-            )
+        elif method == "reduceat":
+            # Combine input mask with where_mask
+            combined_mask = combine_masks((masks[0], where_mask), copy=False)
+            indices = unmasked[1]
+            axis = kwargs.get("axis", 0)
+
+            if combined_mask is False and out_mask is not None:
+                # No elements are masked -> no special handling needed
+                if where_unmasked is True:
+                    out_mask[...] = False
+                else:
+                    # Some partial 'where' usage...
+                    combined_mask = np.broadcast_to(False, unmasked[0].shape)
+
+            if combined_mask is False:
+                # Everything is unmasked, so let NumPy do reduceat directly
+                result = getattr(ufunc, method)(*unmasked, **kwargs)
+                mask = False
+            else:
+                # This means we do have masked elements or partial where
+
+                # Zero-fill or neutral-fill the data array for masked entries
+                data_copy = unmasked[0].copy()
+                data_copy[combined_mask] = 0.0  # or some neutral value for that ufunc
+
+                # Call reduceat on the modified data
+                kwargs["out"] = None  # We'll handle out manually
+                result = getattr(ufunc, method)(data_copy, indices, **kwargs)
+
+                # Build the output mask
+                shape_out = list(data_copy.shape)
+                shape_out[axis] = len(indices)
+                if out_mask is None:
+                    result_mask = np.zeros(shape_out, dtype=bool)
+                else:
+                    result_mask = out_mask
+
+                # For each chunk, check if positions are fully masked
+                for i in range(len(indices)):
```

Comment:
I think you could get most/all of the way with a combination of `np.logical_or(mask, indices)` and `np.logical_and(mask, indices)`, following the logic used in `reduce`.